### PR TITLE
[Qwen3ASR] Add hotword parsing, and fix language parsing and training.

### PR DESCRIPTION
**Pull request by ebezzam**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47111)

# What does this PR do?

A few patches for Qwen3 ASR:
- add support for custom hotwords, e.g. passing custom vocabulary to help transcription like [VibeVoice](https://huggingface.co/microsoft/VibeVoice-ASR-HF#providing-context). Original [model card](https://huggingface.co/Qwen/Qwen3-ASR-1.7B) and [repo](https://github.com/QwenLM/Qwen3-ASR) didn't show such examples... but a user [comment](https://huggingface.co/Qwen/Qwen3-ASR-0.6B-hf/discussions/2) on the new checkpoints brought up this feature not handled by current code. And looking into original code, I see this context [arg](https://github.com/QwenLM/Qwen3-ASR/blob/7c6daf77a2421100f5fb066495372c00129d39ff/qwen_asr/inference/qwen3_asr.py#L303).
- language hint should be handled differently
- while doing above, I noticed a training bug which was copied from other models, because AudioFlamingo3, GlmASR, and MusicFlamingo were affected by [this refactor](https://github.com/huggingface/transformers/pull/45493) (I'll open a separate PR to fix them)

### CODE REVIEW

File: src/transformers/models/qwen3_asr/convert_qwen3_asr_to_hf.py [hunk 1]
```diff
@@ -202,6 +202,62 @@ def clean_config(src_root: Path, model_type: str) -> dict:
 
 
 # fmt: off
+# Extends the original repo's chat template with assistant-turn rendering:
+# - system turn: concatenated text of all system messages (context/hotwords)
+# - user turn: one audio placeholder per audio input
+# - assistant turns: rendered verbatim, wrapped in `{% generation %}` for assistant token masking.
+#   Used to prefill the forced language ("language <NAME><asr_text>") via
+#   `apply_chat_template(..., continue_final_message=True)` and for training targets.
+ASR_CHAT_TEMPLATE = (
```
Comment (ebezzam) [inline src/transformers/models/qwen3_asr/convert_qwen3_asr_to_hf.py:211]: would need to update chat templates of [Qwen/Qwen3-ASR-1.7B-hf](https://huggingface.co/Qwen/Qwen3-ASR-1.7B-hf) and [Qwen/Qwen3-ASR-0.6B-hf](https://huggingface.co/Qwen/Qwen3-ASR-0.6B-hf). I can see with model authors if we do this direction

OR we could go with [this approach](https://github.com/huggingface/transformers/pull/47111/changes#r3529919851) to not have to update the chat template (but perhaps unconventional approach with `apply_chat_template`)

File: src/transformers/models/qwen3_asr/processing_qwen3_asr.py [hunk 2]
```diff
@@ -496,7 +502,8 @@ def apply_transcription_request(
         self,
         audio: AudioInput | list[AudioInput],
         language: str | list[str] | None = None,
-        **kwargs,
+        prompt: str | list[str] | None = None,
```
Comment (ebezzam) [inline src/transformers/models/qwen3_asr/processing_qwen3_asr.py:496]: adding prompt arg like VibeVoice ASR [here](https://github.com/huggingface/transformers/blob/1da9d1d433c301f59027668b25f42d317ec742df/src/transformers/models/vibevoice_asr/processing_vibevoice_asr.py#L178)

File: src/transformers/models/qwen3_asr/processing_qwen3_asr.py [hunk 3]
```diff
@@ -505,39 +512,89 @@ def apply_transcription_request(
...
+            for messages, lang in zip(conversations, languages):
+                prefill = f"language {lang}<asr_text>" if lang is not None else ""
+                messages.append({"role": "assistant", "content": [{"type": "text", "text": prefill}]})
+            return self.apply_chat_template(
+                conversations,
+                tokenize=True,
+                return_dict=True,
+                continue_final_message=True,
+                **kwargs,
+            )
```
Comment (ebezzam) [inline src/transformers/models/qwen3_asr/processing_qwen3_asr.py]: if we push the new chat template, we can call `self.apply_chat_template` with `tokenize=True`

File: src/transformers/models/qwen3_asr/processing_qwen3_asr.py [hunk 4]
```diff
@@ -505,39 +512,89 @@ def apply_transcription_request(
...
+        text = [
+            rendered if lang is None else f"{rendered}language {lang}<asr_text>"
+            for rendered, lang in zip(text, languages)
+        ]
+
+        sampling_rate = kwargs.get(
+            "sampling_rate", kwargs.get("audio_kwargs", {}).get("sampling_rate", self.feature_extractor.sampling_rate)
         )
+        audio_arrays = [load_audio(item, sampling_rate=sampling_rate) for item in audio_items]
+        return self(text=text, audio=audio_arrays, **kwargs)
```
Comment (ebezzam) [inline src/transformers/models/qwen3_asr/processing_qwen3_asr.py]: Otherwise we could leave the chat template as-is, but we wouldn't call `self.apply_chat_template` with `tokenize=True` and need to call the processor `__call__` manually -> this would be the least breaking approach

File: src/transformers/models/qwen3_asr/processing_qwen3_asr.py [hunk 5]
```diff
@@ -505,39 +512,89 @@ def apply_transcription_request(
...
             conversations.append(messages)
 
-        return self.apply_chat_template(
-            conversations,
-            tokenize=True,
-            add_generation_prompt=True,
-            return_dict=True,
-            **kwargs,
+        # Following the original implementation, the language is forced by prefilling the assistant
+        # turn with "language <NAME><asr_text>" so that the model only generates the transcription.
```
Comment (ebezzam) [inline src/transformers/models/qwen3_asr/processing_qwen3_asr.py]: original: https://github.com/QwenLM/Qwen3-ASR/blob/7c6daf77a2421100f5fb066495372c00129d39ff/qwen_asr/inference/qwen3_asr.py#L463

File: src/transformers/models/qwen3_asr/processing_qwen3_asr.py [hunk 6]
```diff
@@ -444,9 +449,10 @@ def __call__(
         model_inputs = super().__call__(audio=audio, text=text, **kwargs)
 
         if output_labels:
-            labels = model_inputs.pop("mm_token_type_ids")
+            mm_token_type_ids = model_inputs.pop("mm_token_type_ids")
+            labels = model_inputs["input_ids"].clone()
+            labels[mm_token_type_ids != 0] = -100  # audio positions
             for token_id in [
```
Comment (ebezzam) [inline src/transformers/models/qwen3_asr/processing_qwen3_asr.py]: bug in preparing labels!

AudioFlamingo3 also affected [here](https://github.com/huggingface/transformers/blob/1da9d1d433c301f59027668b25f42d317ec742df/src/transformers/models/audioflamingo3/processing_audioflamingo3.py#L101), and as a result GlmASR and MusicFlamingo which do modular from this processor.

**Comment by ebezzam** [comment]
run-slow: qwen3_asr

File: docs/source/en/model_doc/qwen3_asr.md [hunk 7]
```diff
@@ -187,9 +209,36 @@ for text in transcriptions:
     print(text)
 ```
 
+The language can be forced through the chat template by *prefilling* the assistant turn with `language <NAME><asr_text>` and passing `continue_final_message=True`, which is what `apply_transcription_request` does under the hood:
```
Comment (vasqu) [inline docs/source/en/model_doc/qwen3_asr.md]: could we kind of combine it with the previous instead with a small comment?

Iiuc then its possible not only to pass to the chat template but also give it as prefix?

File: src/transformers/models/qwen3_asr/convert_qwen3_asr_to_hf.py (hunk 1 above)
Comment (vasqu) [reply src/transformers/models/qwen3_asr/convert_qwen3_asr_to_hf.py:211]: Imo its best to update the template to avoid headaches. This cries workaround an potentially messy maintenance

File: src/transformers/models/qwen3_asr/processing_qwen3_asr.py (hunk 6 above)
Comment (vasqu) [reply src/transformers/models/qwen3_asr/processing_qwen3_asr.py]: yep same comment util worthy!

File: src/transformers/models/qwen3_asr/processing_qwen3_asr.py [hunk 8]
```diff
@@ -505,39 +512,89 @@ def apply_transcription_request(
...
+        elif isinstance(prompt, str):
+            prompts = [prompt] * batch_size
+        elif isinstance(prompt, (list, tuple)):
+            if len(prompt) != batch_size:
+                raise ValueError(
+                    f"Received {len(prompt)} prompt(s) for {batch_size} audio sample(s); counts must match."
+                )
+            prompts = list(prompt)
+        else:
+            raise TypeError("`prompt` must be a string, a sequence of strings, or `None`.")
```
Comment (vasqu) [inline src/transformers/models/qwen3_asr/processing_qwen3_asr.py]: I would check whether something similar already exists? this looks failry standard and should work for most standard primitives

File: src/transformers/models/qwen3_asr/processing_qwen3_asr.py [hunk 9]
```diff
@@ -505,39 +512,89 @@ def apply_transcription_request(
...
+                and the underlying processor call (for example `text_kwargs`, `audio_kwargs`, ...).
 
         Returns:
             [`BatchFeature`]: Processor outputs ready to be passed to
             [`Qwen3ASRForConditionalGeneration.generate`].
         """
-        audio_items = make_list_of_audio_chat_template(audio)
+        audio_items = list(make_list_of_audio_chat_template(audio))
+        if is_torch_available():
+            audio_items = [el.detach().cpu().numpy() if isinstance(el, torch.Tensor) else el for el in audio_items]
```
Comment (vasqu) [inline src/transformers/models/qwen3_asr/processing_qwen3_asr.py]: why is this needed?

File: src/transformers/models/qwen3_asr/processing_qwen3_asr.py (hunk 3 above)
Comment (vasqu) [reply src/transformers/models/qwen3_asr/processing_qwen3_asr.py]: yes please

File: tests/models/qwen3_asr/test_processing_qwen3_asr.py [hunk 10]
```diff
@@ -36,7 +36,7 @@ class Qwen3ASRProcessorTest(ProcessorTesterMixin, unittest.TestCase):
     @classmethod
     @require_torch
     def setUpClass(cls):
-        cls.checkpoint = "Qwen/Qwen3-ASR-0.6B-hf"
+        cls.checkpoint = "bezzam/Qwen3-ASR-0.6B-hf"
```
Comment (vasqu) [inline tests/models/qwen3_asr/test_processing_qwen3_asr.py]: we can open a PR and add a revision for now no? Ideally no revision before merge then

File: tests/models/qwen3_asr/test_processing_qwen3_asr.py [hunk 11]
```diff
@@ -115,6 +115,36 @@ def test_apply_transcription_request_with_language(self):
         for key in ("input_ids", "attention_mask", "input_features", "input_features_mask"):
             self.assertIn(key, outputs)
 
+        # The language is forced by appending "language <NAME><asr_text>" after the generation prompt
+        decoded = processor.tokenizer.decode(outputs["input_ids"][0])
+        self.assertTrue(decoded.endswith("<|im_start|>assistant\nlanguage English<asr_text>"))
+
+    @require_torch
+    def test_apply_transcription_request_with_prompt(self):
```
Comment (vasqu) [inline tests/models/qwen3_asr/test_processing_qwen3_asr.py:123]: What happens if we have both language and prompt? Should they interact in a specific order or should we make it mutually exclusive etc? I think it would make sense to allow both but lan -> prompt

**Review by vasqu — COMMENTED** [review]
Some initial comments. I would nudge towards updating the templates remotely tbh as it will get messy otherwise

File: src/transformers/models/qwen3_asr/processing_qwen3_asr.py (hunk 9 above)
Comment (ebezzam) [reply src/transformers/models/qwen3_asr/processing_qwen3_asr.py]: don't need, copied from vibevoice but we can keep as tensors

File: tests/models/qwen3_asr/test_processing_qwen3_asr.py (hunk 11 above)
Comment (ebezzam) [reply tests/models/qwen3_asr/test_processing_qwen3_asr.py:123]: both are allowed! as seen in the chat template example (input repeated below)
- prompts are in the user role
- language is in the assistant role

their model was trained to have language forcing in the assistant role, so I guess we need to keep it that way?

```
chat_template = [
    [
        # Context/hotwords as system message
        {"role": "system", "content": [{"type": "text", "text": "Vocabulary: Quilter, apostle, gospel."}]},
        {
            "role": "user",
            "content": [
                {
                    "type": "audio",
                    "path": "https://huggingface.co/datasets/bezzam/audio_samples/resolve/main/librispeech_mr_quilter.wav",
                },
            ],
        },
        # empty prefill since forcing language in the other sample
        {"role": "assistant", "content": [{"type": "text", "text": ""}]},
    ],
    [
        {
            "role": "user",
            "content": [
                {
                    "type": "audio",
                    "path": "https://qianwen-res.oss-cn-beijing.aliyuncs.com/Qwen3-ASR-Repo/asr_zh.wav",
                },
            ],
        },
        {"role": "assistant", "content": [{"type": "text", "text": "language Chinese<asr_text>"}]},
    ],
]
```

File: src/transformers/models/qwen3_asr/processing_qwen3_asr.py (hunk 8 above)
Comment (ebezzam) [reply src/transformers/models/qwen3_asr/processing_qwen3_asr.py]: I couldn't find one, but creating `prepare_prompt_input` in audio utils because this was copied from VibeVoice, and future audio LMs may use this!

File: tests/models/qwen3_asr/test_processing_qwen3_asr.py [hunk 12]
```diff
@@ -37,9 +37,10 @@ class Qwen3ASRProcessorTest(ProcessorTesterMixin, unittest.TestCase):
     @require_torch
     def setUpClass(cls):
         cls.checkpoint = "Qwen/Qwen3-ASR-0.6B-hf"
+        cls.revision = "refs/pr/3"  # TODO: set to main after merge
```
Comment (ebezzam) [inline tests/models/qwen3_asr/test_processing_qwen3_asr.py]: @vasqu Opened PRs to the model cards:
- https://huggingface.co/Qwen/Qwen3-ASR-1.7B-hf/discussions/2
- https://huggingface.co/Qwen/Qwen3-ASR-0.6B-hf/discussions/3

When we are ready to merge:
- I will tell model authors to merge
- we set revision to main (or remove altogether cls.revision)

**Comment by ebezzam** [comment]
run-slow: qwen3_asr

File: src/transformers/models/qwen3_asr/processing_qwen3_asr.py [hunk 13]
```diff
@@ -505,37 +503,52 @@ def apply_transcription_request(
...
+            if prompt_text is not None:
+                messages.append({"role": "system", "content": [{"type": "text", "text": prompt_text}]})
             messages.append({"role": "user", "content": [_audio_content_item(audio_item)]})
             conversations.append(messages)
 
+        # The language is forced by prefilling the assistant turn with "language <NAME><asr_text>"
+        # so that the model only generates the transcription text in the desired language
+        for messages, lang in zip(conversations, languages):
+            prefill = f"language {lang}<asr_text>" if lang is not None else ""
+            messages.append({"role": "assistant", "content": [{"type": "text", "text": prefill}]})
```
Comment (vasqu) [inline src/transformers/models/qwen3_asr/processing_qwen3_asr.py:546]: Ah ok one is the user and one is the assistant I think I missed that last time. Really weird template

File: src/transformers/audio_utils.py [hunk 14]
```diff
@@ -426,6 +426,39 @@ def make_list_of_audio_chat_template(
     return make_list_of_audio(audio)
 
 
+def prepare_prompt_input(
```
Comment (vasqu) [inline src/transformers/audio_utils.py]: maybe even processing utils or similar? not really sure if its audio only but super nit

File: tests/models/qwen3_asr/test_processing_qwen3_asr.py
Comment (vasqu) [inline tests/models/qwen3_asr/test_processing_qwen3_asr.py:1]: Lets be careful that the processor is not too heavy for CI?

Comment (vasqu) [reply tests/models/qwen3_asr/test_processing_qwen3_asr.py:1]: We can make a dummy one imo

**Review by vasqu — APPROVED** [review]
LGTM, would be nice to wait for the merge on the hub before merging this one tho 🤗

File: tests/models/qwen3_asr/test_processing_qwen3_asr.py
Comment (ebezzam) [reply tests/models/qwen3_asr/test_processing_qwen3_asr.py:1]: FYI checked with @ydshieh. He's working on a script/workflow to make tiny model copies for the processor tests (e.g. see the `tiny-processor` models [here](https://huggingface.co/hf-internal-testing/models)). So if it passes with the current checkpoint, we'll let him create the tiny processor in a follow-up PR

Comment (vasqu) [reply tests/models/qwen3_asr/test_processing_qwen3_asr.py:1]: Sounds good, feel free to merge then

Comment (ebezzam) [reply tests/models/qwen3_asr/test_processing_qwen3_asr.py:1]: thanks! waiting for the model card PRs to be merged

**Comment by ebezzam** [comment]
run-slow: qwen3_asr
